scripts/common_test.py

```diff
@@ -557,6 +583,91 @@ def mock_get_label(unused_self, unused_name):
                         release_constants.LABEL_FOR_RELEASED_PRS)):
                 common.check_prs_for_current_release_are_released(mock_repo)
 
+    def test_kill_processes_based_on_regex(self):
+        killed = []
+
+        def mock_kill(p):
+            killed.append(MockPsutilProcess.cmdlines[p.index])
+
+        def mock_cmdlines(p):
+            return MockPsutilProcess.cmdlines[p.index]
+
+        def mock_process_iter():
+            return [MockPsutilProcess(0), MockPsutilProcess(1)]
+
+        process_iter_swap = self.swap_with_checks(
+            psutil, 'process_iter', mock_process_iter)
+        kill_swap = self.swap(MockPsutilProcess, 'kill', mock_kill)
+        cmdlines_swap = self.swap(MockPsutilProcess, 'cmdline', mock_cmdlines)
+        with process_iter_swap, kill_swap, cmdlines_swap:
+            common.kill_processes_based_on_regex(r'.*dev_appserver\.py')
+        self.assertEqual(killed, [MockPsutilProcess.cmdlines[0]])
+
+    def test_kill_processes_based_on_regex_when_access_denied(self):
+        killed = []
+
+        def mock_kill(p):
+            killed.append(MockPsutilProcess.cmdlines[p.index])
+
+        def mock_cmdlines(p):
+            if p.index == 0:
+                raise psutil.AccessDenied()
+            return MockPsutilProcess.cmdlines[p.index]
+
+        def mock_process_iter():
+            return [MockPsutilProcess(0), MockPsutilProcess(1)]
+
+        process_iter_swap = self.swap_with_checks(
+            psutil, 'process_iter', mock_process_iter)
+        kill_swap = self.swap(MockPsutilProcess, 'kill', mock_kill)
+        cmdlines_swap = self.swap(MockPsutilProcess, 'cmdline', mock_cmdlines)
+        with process_iter_swap, kill_swap, cmdlines_swap:
+            common.kill_processes_based_on_regex(r'.*dev_appserver\.py')
+        self.assertEqual(killed, [])
+
+    def test_kill_process_when_psutil_not_in_path(self):
+        path_swap = self.swap(sys, 'path', [])
+        def mock_process_iter():
+            return []
+        process_iter_swap = self.swap(psutil, 'process_iter', mock_process_iter)
+        with path_swap, process_iter_swap:
+            common.kill_processes_based_on_regex('')
+
+    def test_inplace_replace_file(self):
+        constant_file = 'constant.js'
+        origin_lines = [
+            '"RANDMON1" : "randomValue1"\n',
+            '"312RANDOM" : "ValueRanDom2"\n',
+            '"DEV_MODE": false\n',
+            '"RAN213DOM" : "raNdoVaLue3"\n'
+        ]
+        expected_lines = [line.replace('\n', '') for line in origin_lines]
+        expected_lines[2] = '"DEV_MODE": true'
+        # pylint: disable=unused-argument
+        def mock_input(files, inplace, backup):
```

Comment:
Don't mock this. Do the operation on a test file. You can use core/tests/data to store test files and expected values.

Make sure to restore things to their original state after the test.

Also please make sure that the tests are comprehensive, in terms of behaviour. E.g. they test regexes with newlines, overlapping matches, etc.

Reply:
I tried to put an actual file, but the problem is that there are many  `PRINT` invoked in other context, which are all dumped into the actual file.

Reply:
I think that's a negative consequence of us using the fileinput API. Let's maybe not do that; let's use 

```
    with open(..., 'r') as f:
        ...
```

and

```
    with open(..., 'w') as f:
        ...
```

instead.